src/app/rosetta/test-agent/operation_expectation.ml

```diff
@@ -27,27 +30,56 @@ let similar t (op : Operation.t) =
         fail err
   in
   let test b err = if b then return () else fail err in
-  let%bind () =
+  let%map _ =
     opt_eq t.amount op.amount ~err:Amount ~f:(fun x y ->
         test String.(equal (string_of_int x) y.Amount.value) Amount )
-  in
-  let%bind () =
+  and _ =
     opt_eq t.account op.account ~err:Account ~f:(fun x y ->
-        let%bind () =
+        let%map _ =
           test
             String.(equal x.Account.pk y.Account_identifier.address)
             Account_pk
-        in
-        let%bind y_token_id =
+        and _ =
           match y.metadata with
-          | Some (`Assoc [("token_id", `String x)]) ->
-              return x
+          | Some (`Assoc [("token_id", `String y)]) ->
+              test
+                String.(equal (Unsigned.UInt64.to_string x.token_id) y)
+                Account_token_id
           | _ ->
               fail Account_token_id
         in
-        test
-          String.(equal (Unsigned.UInt64.to_string x.token_id) y_token_id)
-          Account_token_id )
-  in
-  let%bind () = test String.(equal t.status op.status) Status in
-  test String.(equal t._type op._type) Type
+        () )
+  and _ = test String.(equal t.status op.status) Status
+  and _ = test String.(equal t._type op._type) Type in
+  ()
+
+(** Result of the first expected operation that is wrong. Shows the "closest"
+ * match by selecting the one with minimal error reasons *)
+let assert_similar_operations ~situation ~expected ~actual =
+  List.fold expected ~init:(Result.return ()) ~f:(fun acc t ->
+      let open Result.Let_syntax in
+      let%bind () = acc in
+      let validation, op =
```

Comment:
This allows repeated operations etc. If we want to be stricter, it may be worth building a lazier iterator and carrying the non-matched operations in the accumulator, along the lines of
```ocaml
let rec choose_similar t least_bad_error rest = function
  | [] ->
      (* TODO: This may raise if we are expecting more ops than we get *)
      Error (fst (Option.value_exn least_bad_error))
  | op :: ops -> (
      match similar t op with
      | Ok _ ->
          (* Return the operations with this one removed *)
          Ok (rest @ ops)
      | Error es ->
          let l = List.length es in
          let least_bad_error =
            match least_bad_error with
            | Some (_, l') when l' <= l -> least_bad_error
            | _ -> Some ((op, es), l)
          in
          similar t least_bad_error (op :: rest) ops )
in
List.fold_result ~init:(Result.return actual) expected ~f:(fun actual t ->
  choose_similar t [] actual)
```